File: 1011.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
  bool feasible(vector<int> &weights, int capacity, int days) {
    int current_load = capacity, days_taken = 1;
    for (auto weight : weights) {
      current_load -= weight;
      if (current_load < 0) {
        current_load = capacity - weight;
        days_taken += 1;
      }
    }
    return days_taken <= days;
  }

public:
  int shipWithinDays(vector<int> &weights, int days) {
    int l = -1, r = 0;
    for (auto weight : weights) {
      if (l < weight) {
        l = weight;
      }
      r += weight;
    }
    while (l < r) {
      int m = (r + l) / 2;
      if (feasible(weights, m, days)) {
        r = m;
      } else {
        l = m + 1;
      }
    }
    return l;
  }
};

int main() {}

File: 1011.py
class Solution:
    def shipWithinDays(self, weights: list[int], days: int) -> int:
        l, r = -1, 0
        for weight in weights:
            if weight > l:
                l = weight
            r += weight

        def feasible(weights, capacity, days):
            days_taken = 1
            current_load = capacity
            for weight in weights:
                current_load -= weight
                if current_load < 0:
                    current_load = capacity - weight
                    days_taken += 1
            return days_taken <= days

        while l < r:
            m = (l + r) // 2
            if feasible(weights, m, days):
                r = m
            else:
                l = m + 1
        return l


print(Solution().shipWithinDays([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5))
assert Solution().shipWithinDays([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5) == 15
